horovod/run/run.py

```diff
@@ -0,0 +1,405 @@
+
+# Cached information of horovodrun functions be stored in this directory
+CACHE_FOLDER = os.path.join(os.path.expanduser('~'), '.horovod')
+
+# Cache entries will be stale if they are older than this number of minutes
+CACHE_STALENESS_THRESHOLD_MINUTES = 60
+
+# Maximum number of concurrent worker threads to prevent from over saturation.
+MAX_CONCURRENT_EXECUTIONS = 20
+
+# Number of retries for sshing into the hosts
+SSH_RETRIES = 5
+
+
+def use_cache():
+    """
+    If decorates a function, and if cache_disabled is set, it will store the
+        output of the function if it not None. If a function output is None, the
+    execution will not be cached.
+    :return:
+    """
+
+    def wrap(func):
+        def wrap_f(*args, **kwargs):
+            fn_cache = kwargs.pop('fn_cache')
+            if fn_cache is None:
+                results = func(*args, **kwargs)
+            else:
+                cached_result = fn_cache.get(
+                    (func.__name__, tuple(args[0]), frozenset(kwargs.items())))
+                if cached_result is not None:
+                    return cached_result
+                else:
+                    results = func(*args, **kwargs)
+                    if results is not None:
+                        fn_cache.put(
+                            (func.__name__, tuple(args[0]),
+                             frozenset(kwargs.items())),
+                            results)
+            return results
+
+        return wrap_f
+
+    return wrap
+
+
+@use_cache()
+def _all_hosts_ssh_successful(host_addresses, ssh_port=None):
+    """
+    checks if ssh can successfully be performed to all the hosts.
+    :param host_addresses: list of addresses to ssh into. for example,
+        ['worker-0','worker-1']
+        ['10.11.11.11', '10.11.11.12']
+    :type host_addresses: list(strings)
+    :return: Returns True if all ssh was successful into all the addresses.
+    """
+
+    def exec_command(command):
+        exit_code = 1
+        msg_stdout = ""
+        msg_stderr = ""
+
+        # Try ssh 5 times
+        for i in range(SSH_RETRIES):
+            stdout_w = six.StringIO()
+            stderr_w = six.StringIO()
+            try:
+                exit_code = safe_shell_exec.execute(command,
+                                                    stdout=stdout_w,
+                                                    stderr=stderr_w)
+                if exit_code == 0:
+                    break
+                else:
+                    msg_stderr = stderr_w.getvalue()
+                    msg_stdout = stdout_w.getvalue()
+            finally:
+                stdout_w.close()
+                stderr_w.close()
+        return exit_code, msg_stderr, msg_stdout
+
+    if ssh_port:
+        ssh_port_arg = "-p {ssh_port}".format(ssh_port=ssh_port)
+    else:
+        ssh_port_arg = ""
+
+    ssh_command_format = 'ssh -o StrictHostKeyChecking=no {host} {ssh_port_arg} ls'
```

Comment:
`date` is probably more security-sensitive than `ls`

Reply:
changed.